File: tests/add_image.cc
#include "gtest/gtest.h"

#include "QueryHandlerTester.h"
#include "AthenaConfig.h"

#include <string>
#include <fstream>
#include <iostream>
#include <mutex>
#include <vector>

using namespace athena;
using namespace Jarvis;

std::string singleAddImage(" \
        { \
            \"AddImage\": { \
                \"operations\": [{ \
                    \"width\": 512, \
                    \"type\": \"resize\", \
                    \"height\": 512  \
                }], \
                \"properties\": { \
                    \"name\": \"brain_0\", \
                    \"doctor\": \"Dr. Strange Love\" \
                }, \
                \"format\": \"png\" \
            } \
        } \
    ");

TEST(AddImage, simpleAdd)
{
    std::string addImg;
    addImg += "[" + singleAddImage + "]";

    // int i = system("rm -r simpleAdd_db");
    Graph db("simpleAdd_db", Graph::Create);
    AthenaConfig::init("./addImage-config.json");
    QueryHandler::init();

    std::mutex mu;

    QueryHandler qh_base(&db, &mu);
    QueryHandlerTester query_handler(qh_base);

    protobufs::queryMessage proto_query;
    proto_query.set_json(addImg);

    std::string image;
    std::ifstream file("test_images/brain.png",
                    std::ios::in | std::ios::binary | std::ios::ate);

    image.resize(file.tellg());

    file.seekg(0, std::ios::beg);
    if( !file.read(&image[ 0 ], image.size()))
        std::cout << "error" << std::endl;

    proto_query.add_blobs(image);

    protobufs::queryMessage response;
    query_handler.pq(proto_query, response);

    Json::Reader json_reader;
    Json::Value json_response;

    // std::cout << response.json() << std::endl;
    json_reader.parse(response.json(), json_response);

    EXPECT_EQ(json_response[0]["AddImage"]["status"].asString(), "0");
}

TEST(AddImage, simpleAddx10)
{
    int total_images = 10;
    std::string string_query("[");

    for (int i = 0; i < total_images; ++i) {
        string_query += singleAddImage;
        if (i != total_images - 1)
            string_query += ",";
    }
    string_query += "]";

    // int i = system("rm -r simpleAddx10_db");
    Graph db("simpleAddx10_db", Graph::Create);

    AthenaConfig::init("./addImage-config.json");
    QueryHandler::init();

    std::mutex mu;
    QueryHandler qh_base(&db, &mu);
    QueryHandlerTester query_handler(qh_base);

    protobufs::queryMessage proto_query;
    proto_query.set_json(string_query);

    std::string image;
    std::ifstream file("test_images/brain.png",
                    std::ios::in | std::ios::binary | std::ios::ate);

    image.resize(file.tellg());

    file.seekg(0, std::ios::beg);
    if( !file.read(&image[ 0 ], image.size()))
        std::cout << "error" << std::endl;

    for (int i = 0; i < total_images; ++i) {
        proto_query.add_blobs(image);
    }

    protobufs::queryMessage response;
    query_handler.pq(proto_query, response);

    Json::Reader json_reader;
    Json::Value json_response;

    // std::cout << response.json() << std::endl;
    json_reader.parse(response.json(), json_response);

    for (int i = 0; i < total_images; ++i) {
        EXPECT_EQ(json_response[i]["AddImage"]["status"].asString(), "0");
    }
}

File: tests/python/addImage.py
#! /usr/bin/python
from threading import Thread
import sys
import os
import urllib
import time
import json
import unittest

import athena # Yeah, baby

hostname = "localhost"

class TestStringMethods(unittest.TestCase):

    #Methos to insert one image
    def insertImage(self, props, db):

        imgs_arr = []
        all_queries = []

        fd = open("../test_images/brain.png")
        imgs_arr.append(fd.read())

        # adds some prop
        props["test_case"] = "test_case_prop"

        img_params = {}
        img_params["properties"] = props
        img_params["format"] = "png"
        img_params["log"] = True

        query = {}
        query["AddImage"] = img_params

        all_queries.append(query)

        db.query(all_queries, [imgs_arr])

    def test_addImage(self):
        db = athena.Athena()
        db.connect(hostname)

        all_queries = []
        imgs_arr = []

        number_of_inserts = 2

        for i in range(0,number_of_inserts):
            #Read Brain Image
            fd = open("../test_images/brain.png")
            imgs_arr.append(fd.read())

            collections = {}
            collections["name"] = "BrainScans"

            op_params_resize = {}
            op_params_resize["height"] = 512
            op_params_resize["width"]  = 512
            op_params_resize["type"] = "resize"

            props = {}
            props["name"] = "brain_" + str(i)
            props["doctor"] = "Dr. Strange Love"

            img_params = {}
            img_params["properties"] = props
            # img_params["collections"] = collections
            img_params["operation"] = op_params_resize
            img_params["format"] = "png"
            img_params["log"] = True

            query = {}
            query["AddImage"] = img_params

            all_queries.append(query)

        response, img_array = db.query(all_queries, [imgs_arr])
        # print athena.aux_print_json(response)

        response = json.loads(response)
        self.assertEqual(len(response), number_of_inserts)
        for i in range(0, number_of_inserts):
            self.assertEqual(response[i]["AddImage"]["status"], 0)

    def test_findEntity(self):
        db = athena.Athena()
        db.connect(hostname)

        prefix_name = "fent_brain_"

        for i in range(0,2):
            props = {}
            props["name"] = prefix_name + str(i)
            self.insertImage(props, db)

        all_queries = []

        for i in range(0,2):
            constraints = {}
            constraints["name"] = ["==", prefix_name + str(i)]

            results = {}
            results["list"] = ["name"]

            img_params = {}
            img_params["constraints"] = constraints
            img_params["results"] = results
            img_params["class"] = "AT:IMAGE"

            query = {}
            query["FindEntity"] = img_params

            all_queries.append(query)

        response, img_array = db.query(all_queries)
        # print athena.aux_print_json(response)

        response = json.loads(response)
        self.assertEqual(response[0]["FindEntity"]["status"], 0)
        self.assertEqual(response[1]["FindEntity"]["status"], 0)
        self.assertEqual(response[0]["FindEntity"]["entities"][0]["name"], prefix_name + "0")
        self.assertEqual(response[1]["FindEntity"]["entities"][0]["name"], prefix_name + "1")

    def test_findImage(self):
        db = athena.Athena()
        db.connect(hostname)

        prefix_name = "fimg_brain_"

        for i in range(0,2):
            props = {}
            props["name"] = prefix_name + str(i)
            self.insertImage(props, db)

        all_queries = []

        for i in range(0,2):
            constraints = {}
            constraints["name"] = ["==", prefix_name + str(i)]

            img_params = {}
            img_params["constraints"] = constraints
            img_params["log"] = True

            query = {}
            query["FindImage"] = img_params

            all_queries.append(query)

        response, img_array = db.query(all_queries)
        # print athena.aux_print_json(response)

        response = json.loads(response)
        self.assertEqual(response[0]["FindImage"]["status"], 0)
        self.assertEqual(response[1]["FindImage"]["status"], 0)
        # self.assertEqual(response[0]["FindImage"]["entities"][0]["name"], prefix_name + "0")
        # self.assertEqual(response[1]["FindImage"]["entities"][0]["name"], prefix_name + "1")
        self.assertEqual(len(img_array), 2)

    def test_findImage_multiple_res(self):
        db = athena.Athena()
        db.connect(hostname)

        prefix_name = "fimg_brain_multiple"

        number_of_inserts = 4
        for i in range(0,number_of_inserts):
            props = {}
            props["name"] = prefix_name
            self.insertImage(props, db)

        constraints = {}
        constraints["name"] = ["==", prefix_name]

        results = {}
        results["list"] = ["name"]

        img_params = {}
        # img_params["results"] = results
        img_params["constraints"] = constraints
        img_params["log"] = True

        query = {}
        query["FindImage"] = img_params

        all_queries = []
        all_queries.append(query)

        response, img_array = db.query(all_queries)
        # print athena.aux_print_json(response)

        response = json.loads(response)
        self.assertEqual(len(img_array), number_of_inserts)
        self.assertEqual(response[0]["FindImage"]["status"], 0)
        self.assertEqual(response[0]["FindImage"]["returned"], number_of_inserts)


if __name__ == '__main__':
    unittest.main()